satellite/satellitedb/overlaycache.go

```diff
@@ -671,23 +671,84 @@ func (cache *overlaycache) UpdateAddress(ctx context.Context, info *pb.Node, def
 	return Error.Wrap(tx.Commit())
 }
 
-// UpdateStats a single storagenode's stats in the db
-func (cache *overlaycache) UpdateStats(ctx context.Context, updateReq *overlay.UpdateRequest) (stats *overlay.NodeStats, err error) {
+// BatchUpdateStats updates multiple storagenode's stats in one transaction
+func (cache *overlaycache) BatchUpdateStats(ctx context.Context, updateRequests []*overlay.UpdateRequest, batchSize int) (failed storj.NodeIDList, err error) {
 	defer mon.Task()(&ctx)(&err)
+	if len(updateRequests) == 0 {
+		return failed, nil
+	}
 
-	nodeID := updateReq.NodeID
+	doUpdate := func(updateSlice []*overlay.UpdateRequest) (duf storj.NodeIDList, err error) {
+		appendAll := func() {
+			for _, ur := range updateRequests {
+				duf = append(duf, ur.NodeID)
+			}
+		}
+
+		tx, err := cache.db.Open(ctx)
+		if err != nil {
+			appendAll()
+			return duf, Error.Wrap(err)
+		}
+
+		for _, updateReq := range updateSlice {
+			_, err := updateStats(ctx, updateReq, tx)
+			if err != nil {
+				appendAll()
+				return duf, errs.Combine(err, tx.Rollback())
+			}
+		}
+		return duf, Error.Wrap(tx.Commit())
+	}
 
+	var errlist errs.Group
+	length := len(updateRequests)
+	for i := 0; i < length; i += batchSize {
+		end := i + batchSize
+		if end > length {
+			end = length
+		}
+
+		failedBatch, err := doUpdate(updateRequests[i:end])
+		if err != nil && len(failedBatch) > 0 {
+			for _, fb := range failedBatch {
+				errlist.Add(err)
+				failed = append(failed, fb)
+			}
+		}
+	}
+	return failed, errlist.Err()
+}
+
+// UpdateStats a single storagenode's stats in the db
+func (cache *overlaycache) UpdateStats(ctx context.Context, updateReq *overlay.UpdateRequest) (stats *overlay.NodeStats, err error) {
+	defer mon.Task()(&ctx)(&err)
 	tx, err := cache.db.Open(ctx)
 	if err != nil {
 		return nil, Error.Wrap(err)
 	}
+
+	nodeStats, err := updateStats(ctx, updateReq, tx)
+	if err != nil {
+		return nodeStats, errs.Combine(err, tx.Rollback())
+	}
+
+	return nodeStats, errs.Combine(err, tx.Commit())
+}
+
+// UpdateStats a single storagenode's stats in the db
+func updateStats(ctx context.Context, updateReq *overlay.UpdateRequest, tx *dbx.Tx) (stats *overlay.NodeStats, err error) {
```

Comment:
I wonder whether this would end up problematic with multiple servers updating at the same time.

Reply:
LWW (last write wins) would be the guarantee this provides wouldn't it?